restler/utils/logging/trace_db.py

```diff
@@ -0,0 +1,225 @@
+            If fewer messages are available, it will save all available messages.
+        """
+        message = self._log_queue.get()
+        if message is None:
+            # End of tracing signal
+            return
+        self.storage_writer.save(message)
+
+    def get_sequence_id(self):
+        tls_trace_log = get_trace_log()
+        if tls_trace_log is None:
+            return None
+        return tls_trace_log.sequence_id
+
+    def log_request_response(self, request=None, response=None,  tags={}, timestamp=None):
+        if request is None and response is None:
+            raise Exception("ERROR: Request or response must be specified.")
+        try:
+            tls_trace_log = get_trace_log()
+            if tls_trace_log is None:
+                # Logging requests outside of sequence context (e.g. GC or checker requests
+                # that are not part of a sequence)
+                trace_log = RequestTraceLog()
+            else:
+                trace_log = RequestTraceLog(request_id=tls_trace_log.request_id,
+                                              sequence_id=tls_trace_log.sequence_id,
+                                              combination_id=tls_trace_log.combination_id,
+                                              tags=tls_trace_log.tags)
+            if request:
+                trace_log.request = request
+                if timestamp is not None:
+                    trace_log.sent_timestamp = timestamp
+            if response:
+                trace_log.response = response
+                if timestamp is not None:
+                    trace_log.received_timestamp = timestamp
+            if tags:
+                trace_log.tags.extend(tags)
+            self.log(trace_log.to_dict())
+        except Exception as error:
+            # print the callstack
+            import traceback
+            traceback.print_exc()
+            print(f"Warning: Exception logging request/response: {error}")
+            pass
+
+    def load_trace_data(self):
+        self.storage_writer.load()
+
+    def normalize_trace_data(self):
+        self.storage_writer.normalize()
+
+def set_up_custom_serializer(serializer_file_path):
+    """ Set up a custom serializer for the trace database.
+    The custom serializer must inherit from the 'TraceLogWriterBase' class, and accept one argument,
+    which is the settings object specified by the user in the engine settings.
+    """
+    obj = None
+    custom_serializer = import_utilities.import_subclass(serializer_file_path, TraceLogWriterBase)
+    if custom_serializer is not None:
+        # Instantiate using the settings object specified by the user.
+        obj = custom_serializer(Settings().trace_db_serializer_settings)
+    return obj
+
+def set_up_trace_database(root_dir_path=None):
+    global _db
+    if _db is not None:
+        raise Exception("Trace database already initialized")
+
+    if root_dir_path is None:
+        root_dir_path = os.getcwd()
```

Comment:
I wanted to confirm if this was intentional, I noticed that by default the trace_db file is not created in the same directory as the per-experiment logs.

I am seeing trace_data.ndjson created in the Fuzz directory, whereas the network logs are in Fuzz/RestlerResults/experimentXYZ/logs

Reply:
It was intentional, but thinking about this again, it would be cleaner to put this under the experiment directory by default.

Reply:
Moved to ExperimentXYZ